cmd/libsnap-confine-private/cgroup-support-test.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "cgroup-support.c"
+
+#include <fcntl.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "../libsnap-confine-private/cleanup-funcs.h"
+#include "../libsnap-confine-private/test-utils.h"
+#include "cgroup-support.h"
+
+static void sc_set_self_cgroup_path(const char *mock);
+
+static void sc_set_cgroup_root(const char *mock) { cgroup_dir = mock; }
+
+typedef struct _cgroupv2_is_tracking_fixture {
+    char *p;
+    char *root;
+} cgroupv2_is_tracking_fixture;
+
+static void cgroupv2_is_tracking_set_up(cgroupv2_is_tracking_fixture *fixture, gconstpointer user_data) {
+    GError *err = NULL;
+    int fd = g_file_open_tmp("s-c-unit-is-tracking-self-group.XXXXXX", &fixture->p, &err);
+    g_assert_no_error(err);
+    g_close(fd, &err);
+    g_assert_no_error(err);
+    sc_set_self_cgroup_path(fixture->p);
+
+    fixture->root = g_dir_make_tmp("s-c-unit-test-root.XXXXXX", &err);
+    sc_set_cgroup_root(fixture->root);
+}
+
+static void cgroupv2_is_tracking_tear_down(cgroupv2_is_tracking_fixture *fixture, gconstpointer user_data) {
+    GError *err = NULL;
+
+    sc_set_self_cgroup_path("/proc/self/cgroup");
+    g_remove(fixture->p);
+    g_free(fixture->p);
+
+    sc_set_cgroup_root("/sys/fs/cgroup");
+    char *cmd = g_strdup_printf("rm -rf %s", fixture->root);
```

Comment:
It's slightly sad that there is *nothing* like shutil.rmtree() or os.RemoveAll() in glib/glibc or things like this

Reply:
It's probably too late for this, but IMHO it would make sense to write unit tests for the C parts in a higher level language (C++ being the most obvious choice, but I would also consider pytest + `ctypes` here).